### ISSUE TITLE
Feature Request: Ignore coverage by environment variables

### ISSUE THREAD

**Issue by yatki** [istanbuljs/nyc#969]
## Motivation
I have different types of tests like unit, integration and functional. When I write my unit tests for some functions that only does I/O operations and I'd like to write only integration tests for those. But this will cause decrease in my unit test code coverage.

I'd like to tell istanbul to ignore my code block depending on the test type it's running. 

## Example

> A coverage skip hint looks like /* istanbul ignore \<word\>[non-word] [optional-docs] */

Respecting the description above; maybe we can have an entry in .nycrc to specify custom words and look for an evironment variable:

```json
{
  "nyc": {
    "check-coverage": true,
    "words": [
      { "word": "unitAndFn", "nycIgnore": ["unit", "functional" ] },
      { "word": "integration", "nycIgnore": ["integration" ] }
    ]
  }
}
``` 
 And in my source code I can use;
```js
/* A code piece that doesn't need unit test coverage*/
/* istanbul ignore unitAndFn  */
const functionA = () => {
   // ...
};

/* A code piece that doesn't need integration test coverage*/
/* istanbul ignore integration  */
const functionB = () => {
   // ...
};
```
And when running the tests:
```
# contains unit so unitAndFn word will be used
~ $ NYC_IGNORE=unit nyc jest 

# contains integration so integration word will be used
~ $ NYC_IGNORE=integration nyc jest
```
This is just my initial proposal. If you like the idea, we can, of course, find a better way to implement it. Maybe accepting environment variables as an array is too much. I just thought maybe gives more flexibility. I'm happy to elaborate the idea.

And maybe there is already a way to do this? I'd be happy if you could let me know if there is.

# Pitch

- I think this will help developers, first of all, remember which functionalities were covered with which test suites they have by looking at comments.
- Also, this will help developers to have more meaningful code coverage reports in their relative test suite reports.

**Comment by coreyfarrell** [comment]
I think this would need to start with expanded functionality in `istanbul-lib-instrument` in the [base library monorepo](https://github.com/istanbuljs/istanbuljs/).  I think at that level the instrumenter should not look at environmental variables, it should take API options.  In addition to what you've suggested I think the ability to ignore coverage based on `process.platform` would be useful.

I do think the syntax of the istanbul hints needs some more thought.  We need to ensure the new syntax doesn't conflict with existing hints (`istanbul ignore next`, `istanbul ignore else`, etc).

**Comment by coreyfarrell** [comment]
A possible hint pattern for platform based filtering:

```js
/* istanbul not-platform win32 ignore next */
if (process.platform === 'win32') {
  console.log('this is only reachable on win32.');
}
/* istanbul platform win32 ignore next */
if (process.platform !== 'win32') {
  console.log('this is only reachable outside win32');
}
```

As for environment testing maybe we could keep this simpler by supporting an `ISTANBUL_ENV` environmental variable.  This variable could be set to a space separated list of identifiers to be checked as below.  This would require some special care to ensure that ISTANBUL_ENV is considered as part of babel and nyc caching.
```js
/* istanbul not-env integration ignore next */
function onlyHitByIntegrationTests() {
}

/* istanbul env integration ignore next */
function onlyHitOutsideIntegrationTests() {
}
```

CC @bcoe @JaKXz any opinions you want to share on this issue?

**Comment by yatki** [comment]
@coreyfarrell Thank you very much for elaborating the idea. This syntax looks easier to understand and remember.

I'm just thinking if it would be useful to support multiple ignore conditions. 

They can be defined as comma-separated env values like below:

```js
/* istanbul env integration,custom-case ignore next */
function onlyHitOutsideIntegrationTests() {
}
```

or chained comments like below:

```js
/* istanbul env custom-case ignore next */
/* istanbul env integration ignore next */
function onlyHitOutsideIntegrationTests() {
}
```

Comments can be chained with an "or" statement and if any of the comment statements return true, the next block will be ignored. This may allow users to have better control over the coverage. 

Do you guys think, this is a good idea?

**Comment by coreyfarrell** [comment]
I'd rather keep the syntax of individual hints as simple as possible.  I'd prefer we not support testing multiple env's in a single hint.  I'm definitely not in favor of the 'or' syntax.  If multiple comments can be easily supported by the instrumenter I wouldn't have a problem with that.

Really my idea is that you would set multiple values to `ISTANBUL_ENV`, such as `cross-env ISTANBUL_ENV="integration custom-case" nyc ...`, then you could use `custom-case` for certain hints, `integration` for other hints.

**Comment by yatki** [comment]
I like this approach and agree that having comma separated values in a single hint is not a neat solution. Let's ignore that.

Also just to clarify, I didn't mean to write `or` statements inside comments. 
I was just thinking when evaluating multiple comments, the logic between comments would be an "or" condition. Sorry for not being clear, let me give an example.

```js
/* istanbul env custom-case ignore next */
/* istanbul env integration ignore next */
function onlyIgnoredIfEnvIsIntegrationOrCustomCase() {
}
```
So in the example above, the code block would be ignored only if env is `custom-case` **or** `integration`. So actually, at the end, each hint will be evaluated individually and when multiple hints were true, the first one will be evaluated and the next hints will be ignored. 

So, if `cross-env ISTANBUL_ENV="integration custom-case" nyc ...` command was executed, `/* istanbul env custom-case ignore next */` statement will be evaluated first and the hints coming after that will be ignored.

Lastly, I'm already satisfied with having a single hint for a statement. This would solve all of my problems at the moment. I'm kind of brainstorming at this point. if having multiple comments wouldn't bring too much complexity to the instrumenter, maybe it's a feature worth having, maybe it's not.

**Comment by coreyfarrell** [comment]
I've labeled this ticket so the stale bot will not touch it again.  I like some of the ideas here, just don't have the cycles to work on it currently.

**Comment by coreyfarrell** [comment]
FWIW with nyc 14+ you can create `nyc.config.js` instead of `.nycrc.json`.  This allows logic to be used which can allow conditional exclude functionality.  For example see [test-exclude](https://github.com/istanbuljs/istanbuljs/tree/master/packages/test-exclude), in particular the way that nyc.config.js deals with `is-outside-dir*.js`.  I still feel it would be nice to be able to deal with situations like this using `istanbul ignore` comments but wanted to mention this as a potential work-around.

**Comment by yatki** [comment]
@coreyfarrell Thank you very much for the info. Sure this would also help a lot at the moment.

I also agree that having comments would be a really nice feature.

**Comment by Ryuno-Ki** [comment]
Not sure, whether I got the request correctly …
But in the Python world, you could use [pytest with custom markers](https://docs.pytest.org/en/latest/example/markers.html) to control the test execution (and via `pytest-cov`, the coverage) to a certain degree …

**Comment by yatki** [comment]
@Ryuno-Ki Correct me if I'm wrong but I think this is more like tagging the test suites, and filtering which tests will be executed. Eg: [jest --testNamePattern](https://jestjs.io/docs/en/cli#testnamepattern-regex) or [mocha --grep](https://mochajs.org/#-grep-regexp-g-regexp).

What we are talking about is to have separate code coverage report for unit tests and integration tests.

> I have different types of tests like unit, integration and functional. For some functions that only does I/O operations, I'd like to write **only** integration tests and skip writing unit tests. But this will decrease my **unit test** code coverage.

**Comment by Ryuno-Ki** [comment]
Okay, how does the test runner know, whether this is an unit or integration test?
I was thinking about marking them manually.
That's why I came up with my comment.

**Comment by yatki** [comment]
You are not wrong. I was refering to the link you gave but, yes basically we are talking about the same thing I guess: Having a comment pattern like
`/* istanbul env integration ignore next */` on top of code blocks to manually mark them :)

> Okay, how does the test runner know, whether this is an unit or integration test?

It doesn't. We just place unit tests, and integration tests in separate folders and tell to the test runner to run tests under the corresponding folder :)

**Comment by d-tw** [comment]
Hi,

I also have different test-suites (unit, integration, contract, etc.), which don't all have the same coverage (typically integration and contract are more or less a subset of unit).

In my team, our CI is set to fail if coverage is less than 100% (that doesn't mean that every single line of code is tested, nor is that the objective, it just means that code that isn't tested is intentionally omitted via the correct ignore comments).

Is there currently any way to conditionally ignore blocks of code from coverage? (blocks, not files)
